hazelcast/src/main/java/com/hazelcast/internal/networking/nio/NioNetworking.java

```diff
@@ -209,20 +209,18 @@ public void restart() {
             thread.id = i;
             thread.setSelectorWorkaroundTest(selectorWorkaroundTest);
             outThreads[i] = thread;
-            metricsRegistry.newProbeBuilder("tcp.outputThread")
-                           .withTag("thread", thread.getName())
-                           .scanAndRegister(thread);
             thread.start();
         }
         this.outputThreads = outThreads;
 
         startIOBalancer();
+
+        metricsRegistry.registerDynamicMetricsProvider(this);
```

Comment:
Shall we register metrics before we start `thread`?

Reply:
Not necessarily. Here we just register the root of the metrics that will be dynamically and periodically collected later on. I improved `provideDynamicMetrics` in this class with a null-check for `ioBalancer` to be on the safe side.